File: backend/app/schemas/fight.py
from pydantic import BaseModel
from typing import Optional, Dict, Any, List
from datetime import datetime
import uuid


class FightBase(BaseModel):
    event_name: str
    date: datetime
    fighter_a_id: uuid.UUID
    fighter_b_id: uuid.UUID
    weight_class: str
    is_main_event: bool = False
    is_title_fight: bool = False
    odds: Dict[str, float] = {}
    winner_id: Optional[uuid.UUID] = None
    method: Optional[str] = None
    round: Optional[int] = None
    time: Optional[str] = None
    ml_insights: Dict[str, Any] = {}
    is_completed: bool = False


class FightCreate(FightBase):
    pass


class FightUpdate(BaseModel):
    event_name: Optional[str] = None
    date: Optional[datetime] = None
    odds: Optional[Dict[str, float]] = None
    winner_id: Optional[uuid.UUID] = None
    method: Optional[str] = None
    round: Optional[int] = None
    time: Optional[str] = None
    ml_insights: Optional[Dict[str, Any]] = None
    is_completed: Optional[bool] = None


class FightInDB(FightBase):
    id: uuid.UUID
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class Fight(FightInDB):
    pass


class FightWithFighters(Fight):
    fighter_a: "Fighter"
    fighter_b: "Fighter"
    winner: Optional["Fighter"] = None


class PredictionBase(BaseModel):
    fight_id: uuid.UUID
    predicted_winner: uuid.UUID
    predicted_method: str
    predicted_round: Optional[int] = None
    confidence: int = 50


class PredictionCreate(PredictionBase):
    pass


class PredictionUpdate(BaseModel):
    predicted_winner: Optional[uuid.UUID] = None
    predicted_method: Optional[str] = None
    predicted_round: Optional[int] = None
    confidence: Optional[int] = None


class PredictionInDB(PredictionBase):
    id: uuid.UUID
    user_id: uuid.UUID
    actual_result: Optional[str] = None
    elo_change: int = 0
    points_earned: int = 0
    is_correct: Optional[bool] = None
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class Prediction(PredictionInDB):
    pass


class PredictionWithFight(Prediction):
    fight: Fight
    predicted_winner_fighter: "Fighter" 

File: frontend/lib/models/fight.dart
import 'package:json_annotation/json_annotation.dart';
import 'fighter.dart';

part 'fight.g.dart';

@JsonSerializable()
class Fight {
  final String id;
  final String eventName;
  final DateTime date;
  final String fighterAId;
  final String fighterBId;
  final String weightClass;
  final bool isMainEvent;
  final bool isTitleFight;
  final Map<String, double> odds;
  final String? winnerId;
  final String? method;
  final int? round;
  final String? time;
  final Map<String, dynamic> mlInsights;
  final bool isCompleted;
  final DateTime createdAt;
  final DateTime? updatedAt;
  
  // Related data (not from API)
  final Fighter? fighterA;
  final Fighter? fighterB;
  final Fighter? winner;

  const Fight({
    required this.id,
    required this.eventName,
    required this.date,
    required this.fighterAId,
    required this.fighterBId,
    required this.weightClass,
    required this.isMainEvent,
    required this.isTitleFight,
    required this.odds,
    this.winnerId,
    this.method,
    this.round,
    this.time,
    required this.mlInsights,
    required this.isCompleted,
    required this.createdAt,
    this.updatedAt,
    this.fighterA,
    this.fighterB,
    this.winner,
  });

  factory Fight.fromJson(Map<String, dynamic> json) => _$FightFromJson(json);
  Map<String, dynamic> toJson() => _$FightToJson(this);

  Fight copyWith({
    String? id,
    String? eventName,
    DateTime? date,
    String? fighterAId,
    String? fighterBId,
    String? weightClass,
    bool? isMainEvent,
    bool? isTitleFight,
    Map<String, double>? odds,
    String? winnerId,
    String? method,
    int? round,
    String? time,
    Map<String, dynamic>? mlInsights,
    bool? isCompleted,
    DateTime? createdAt,
    DateTime? updatedAt,
    Fighter? fighterA,
    Fighter? fighterB,
    Fighter? winner,
  }) {
    return Fight(
      id: id ?? this.id,
      eventName: eventName ?? this.eventName,
      date: date ?? this.date,
      fighterAId: fighterAId ?? this.fighterAId,
      fighterBId: fighterBId ?? this.fighterBId,
      weightClass: weightClass ?? this.weightClass,
      isMainEvent: isMainEvent ?? this.isMainEvent,
      isTitleFight: isTitleFight ?? this.isTitleFight,
      odds: odds ?? this.odds,
      winnerId: winnerId ?? this.winnerId,
      method: method ?? this.method,
      round: round ?? this.round,
      time: time ?? this.time,
      mlInsights: mlInsights ?? this.mlInsights,
      isCompleted: isCompleted ?? this.isCompleted,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      fighterA: fighterA ?? this.fighterA,
      fighterB: fighterB ?? this.fighterB,
      winner: winner ?? this.winner,
    );
  }

  bool get isUpcoming => date.isAfter(DateTime.now());
  bool get isPast => date.isBefore(DateTime.now());

  @override
  String toString() {
    return 'Fight(id: $id, eventName: $eventName, date: $date)';
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is Fight && other.id == id;
  }

  @override
  int get hashCode => id.hashCode;
} 